proposals/lip-0049.md

```diff
@@ -223,135 +223,171 @@ When processing a cross-chain message `ccm`, the functions below may be called b
+        Token.transferBetweenEscrows(ccm.sendingChainID, ccm.receivingChainID, feeTokenID, fee)
+
         addToOutbox(ccm.receivingChainID, ccm)
+        # Emit CCM forwarded event.
+        emitEvent(
+            moduleID = MODULE_ID_INTEROPERABILITY,
+            typeID = TYPE_ID_CCM_PROCESSED,
+            data = {"ccmID": ccmID, "status": CCM_PROCESSED_STATUS_FORWARDED},
+            topics = [ccm.sendingChainID, ccm.receivingChainID]
+        )
         return
 
+    # If ccm status is not ok, stop the processing of this message
     if ccm.status != CCM_STATUS_OK:
-        # Stop the processing of this message
+        # Emit CCM discarded event.
+        emitEvent(
+            moduleID = MODULE_ID_INTEROPERABILITY,
+            typeID = TYPE_ID_CCM_PROCESSED,
+            data = {"ccmID": ccmID, "status": CCM_PROCESSED_STATUS_DISCARDED},
+            topics = [ccm.sendingChainID, ccm.receivingChainID]
+        )
         return
 
     # In all following cases, bounce the message with updated status
     # Note: the fee was not assigned to the mainchain relayer and is also bounced
-    bounce(ccm)
+    
+    # Emit CCM processed event: channel unavailable.
+    emitEvent(
+        moduleID = MODULE_ID_INTEROPERABILITY,
+        typeID = TYPE_ID_CCM_PROCESSED,
+        data = {"ccmID": ccmID, "status": CCM_PROCESSED_STATUS_CHANNEL_UNAVAILABLE},
+        topics = [ccm.sendingChainID, ccm.receivingChainID]
+    )
+    ccm.status = CCM_STATUS_CHANNEL_UNAVAILABLE
+    addToOutbox(ccm.sendingChainID, ccm)
 
-    # If the chain does not exist or is not yet active don't continue
-    if (chainAccount(ccm.receivingChainID) does not exist
-        or chainAccount(ccm.receivingChainID).status == CHAIN_REGISTERED):  
+    # If the chain does not exist or is not yet active do not continue
+    if chainAccount(ccm.receivingChainID) does not exist
+        or chainAccount(ccm.receivingChainID).status == CHAIN_STATUS_REGISTERED:  
         return
 
     # In the remaining case, the receiving chain account existed and was either
     # terminated or not live
     # If the receiving chain is still active, then it gets terminated
-    if chainAccount(ccm.receivingChainID).status == CHAIN_ACTIVE:
-        terminateChainInternal(ccm.receivingChainID)
+    if chainAccount(ccm.receivingChainID).status == CHAIN_STATUS_ACTIVE:
+        terminateChain(ccm.receivingChainID)
 
     # A sidechain terminated message is returned to the sending chain
     stmParams = {
         "chainID": ccm.receivingChainID,
         "stateRoot": chainAccount(ccm.receivingChainID).lastCertificate.stateRoot
-    } serialized using sidechainTerminatedCCMParamsSchema
+    }
 
-    sendInternal(
+    send(
         EMPTY_FEE_ADDRESS,
         MODULE_ID_INTEROPERABILITY,
-        CROSS_CHAIN_COMMAND_ID_SIDECHAIN_TERMINATED,
+        CROSS_CHAIN_COMMAND_ID_SIDECHAIN_STATUS_TERMINATED,
         ccm.sendingChainID,
         0,
         CCM_STATUS_OK,
         stmParams
```

Comment:
Strictly speaking, I think the params need to be serialized.

Reply:
it is serialized in `send`

Reply:
Ah okay, my bet.